_posts/2021-12-22-federated-learning-candig.md

```diff
@@ -0,0 +1,154 @@
+            print(f"Saving round {rnd} aggregated_weights...")
+            np.savez(f"round-{rnd}-weights.npz", *aggregated_weights)
+        return aggregated_weights
+```
+
+We also need to provide our Strategy class an evaluation function to update server-side model weights and provide evaluation metrics. This function receives the aggregated weights as a parameter and is called once per training round. We also specify a simple config function that is called once per round to tell clients which training round is taking place. We can also send clients other variables here.
+
+```python
+def get_eval_fn(model: LogisticRegression):
+    """Return an evaluation function for server-side evaluation."""
+
+    # Load test data here to avoid the overhead of doing it in `evaluate` itself
+    _, (X_test, y_test) = utils.load_data()
+
+    # The `evaluate` function will be called after every round
+    def evaluate(parameters: fl.common.Weights):
+        # Update model with the latest parameters
+        utils.set_model_params(model, parameters)
+        loss = log_loss(y_test, model.predict_proba(X_test))
+        accuracy = model.score(X_test, y_test)
+        auc_score = roc_auc_score(y_test, model.predict(X_test))
+        return loss, {"accuracy": accuracy, "auc score": auc_score}
+
+    return evaluate
+
+def fit_round(rnd: int) -> Dict:
+    """Send round number to client."""
+    return {"rnd": rnd}
+```
+
+We pass the new `SaveModelStrategy` class these functions on construction before starting the server. We also pass it the minimum amount of connected clients it should have before beginning federated learning.
+
+```python
+model = LogisticRegression()
+    utils.set_initial_params(model)
+    strategy = SaveModelStrategy(
+        min_available_clients=2,
+        eval_fn=get_eval_fn(model),
+        on_fit_config_fn=fit_round,
+    )
+    fl.server.start_server("0.0.0.0:8080", strategy=strategy, config={"num_rounds": 100})
+```
+
+And that's it! Our `utils.load_data` function is called both client and server side whenever necessary to fetch and preprocess training/test data from the aforementioned GraphQL interface, and model weights are saved server-side every 10 rounds for the proposed researcher's convenience.
+
+While there are several other considerations and processes (particularly in security and user accessibility) that are required to make this learning service anywhere near production-ready, we can indeed see that federated learning can integrate easily with CanDIG's preexisting microservices, and that Flower as a framework is flexible and simple to set up standard FL workflows that are relevant to CanDIG.
+
+<!-- links -->
+[Federated Averaging]: https://arxiv.org/pdf/1602.05629.pdf
+[Tensorflow Federated]: https://openid.net/
+[NVFlare]: https://www.openpolicyagent.org/
+[Flower]: https://flower.dev/
+[GraphQL Interface]: https://www.distributedgenomics.ca/posts/secure-cross-service-graphql-interface/
```

Comment:
I think this should be `[Secure Cross-service GraphQL Interface]`

Reply:
Actually - maybe both, GraphQL Interface and Secure Cross-service GraphQL interface (there's two links in the body)